File: MCPForUnity/Editor/Tools/ManageLighting.cs
using System;
using MCPForUnity.Editor.Helpers;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEditor.SceneManagement;
using UnityEngine;
using UnityEngine.Rendering;

namespace MCPForUnity.Editor.Tools
{
    /// <summary>
    /// Tool for managing Unity Lighting:
    /// - Create and configure lights (Directional, Point, Spot, Area)
    /// - Set ambient lighting and environment settings
    /// - Configure fog
    /// - Create and bake reflection probes
    /// - Bake lightmaps
    /// </summary>
    [McpForUnityTool("manage_lighting")]
    public static class ManageLighting
    {
        public static object HandleCommand(JObject @params)
        {
            if (@params == null)
            {
                return new ErrorResponse("Parameters cannot be null.");
            }

            string action = ParamCoercion.CoerceString(@params["action"], null)?.ToLowerInvariant();
            if (string.IsNullOrEmpty(action))
            {
                return new ErrorResponse("'action' parameter is required.");
            }

            try
            {
                return action switch
                {
                    "light_create" => LightCreate(@params),
                    "light_configure" => LightConfigure(@params),
                    "light_get_info" => LightGetInfo(@params),
                    "ambient_set" => AmbientSet(@params),
                    "ambient_get" => AmbientGet(@params),
                    "fog_set" => FogSet(@params),
                    "fog_get" => FogGet(@params),
                    "reflection_probe_create" => ReflectionProbeCreate(@params),
                    "reflection_probe_bake" => ReflectionProbeBake(@params),
                    "lightmap_bake" => LightmapBake(@params),
                    "lighting_get_settings" => LightingGetSettings(@params),
                    _ => new ErrorResponse($"Unknown action: '{action}'. Supported: light_create, light_configure, light_get_info, ambient_set, ambient_get, fog_set, fog_get, reflection_probe_create, reflection_probe_bake, lightmap_bake, lighting_get_settings")
                };
            }
            catch (Exception e)
            {
                McpLog.Error($"[ManageLighting] Action '{action}' failed: {e}");
                return new ErrorResponse($"Error processing action '{action}': {e.Message}");
            }
        }

        #region Light Creation and Configuration

        private static object LightCreate(JObject @params)
        {
            string lightTypeStr = ParamCoercion.CoerceString(@params["lightType"], "Point");
            if (!Enum.TryParse<LightType>(lightTypeStr, true, out var lightType))
            {
                return new ErrorResponse($"Unknown light type: '{lightTypeStr}'. Supported: Directional, Point, Spot, Area");
            }

            Vector3 position = ParseVector3(@params["position"], Vector3.zero);
            Vector3 rotation = ParseVector3(@params["rotation"], lightType == LightType.Directional ? new Vector3(50, -30, 0) : Vector3.zero);

            GameObject lightGo = new GameObject($"{lightType} Light");
            lightGo.transform.position = position;
            lightGo.transform.eulerAngles = rotation;

            Light light = lightGo.AddComponent<Light>();
            light.type = lightType;

            // Apply properties
            ApplyLightProperties(light, @params);

            Undo.RegisterCreatedObjectUndo(lightGo, "Create Light");
            EditorSceneManager.MarkSceneDirty(lightGo.scene);

            return new SuccessResponse($"{lightType} light created.", new
            {
                instanceID = lightGo.GetInstanceID(),
                name = lightGo.name,
                lightType = lightType.ToString(),
                position = new { x = position.x, y = position.y, z = position.z },
                color = ColorToHex(light.color),
                intensity = light.intensity
            });
        }

        private static object LightConfigure(JObject @params)
        {
            Light light = FindLight(@params);
            if (light == null) return LightNotFoundError(@params);

            Undo.RecordObject(light, "Configure Light");
            ApplyLightProperties(light, @params);

            EditorUtility.SetDirty(light);
            EditorSceneManager.MarkSceneDirty(light.gameObject.scene);

            return new SuccessResponse($"Light '{light.name}' configured.", new
            {
                instanceID = light.gameObject.GetInstanceID(),
                lightType = light.type.ToString(),
                color = ColorToHex(light.color),
                intensity = light.intensity,
                range = light.range,
                shadows = light.shadows.ToString()
            });
        }

        private static object LightGetInfo(JObject @params)
        {
            Light light = FindLight(@params);
            if (light == null) return LightNotFoundError(@params);

            return new SuccessResponse($"Light info for '{light.name}'.", new
            {
                instanceID = light.gameObject.GetInstanceID(),
                name = light.name,
                lightType = light.type.ToString(),
                color = ColorToHex(light.color),
                intensity = light.intensity,
                range = light.range,
                spotAngle = light.spotAngle,
                innerSpotAngle = light.innerSpotAngle,
                shadows = light.shadows.ToString(),
                shadowStrength = light.shadowStrength,
                cullingMask = light.cullingMask,
                enabled = light.enabled
            });
        }

        private static void ApplyLightProperties(Light light, JObject @params)
        {
            if (@params["color"] != null)
                light.color = ParseColor(@params["color"], light.color);

            if (@params["intensity"] != null)
                light.intensity = ParamCoercion.CoerceFloat(@params["intensity"], light.intensity);

            if (@params["range"] != null)
                light.range = ParamCoercion.CoerceFloat(@params["range"], light.range);

            if (@params["spotAngle"] != null)
                light.spotAngle = ParamCoercion.CoerceFloat(@params["spotAngle"], light.spotAngle);

            if (@params["innerSpotAngle"] != null)
                light.innerSpotAngle = ParamCoercion.CoerceFloat(@params["innerSpotAngle"], light.innerSpotAngle);

            if (@params["shadows"] != null)
            {
                string shadowStr = ParamCoercion.CoerceString(@params["shadows"], "None");
                if (Enum.TryParse<LightShadows>(shadowStr, true, out var shadowType))
                    light.shadows = shadowType;
            }

            if (@params["shadowStrength"] != null)
                light.shadowStrength = ParamCoercion.CoerceFloat(@params["shadowStrength"], light.shadowStrength);

            if (@params["shadowBias"] != null)
                light.shadowBias = ParamCoercion.CoerceFloat(@params["shadowBias"], light.shadowBias);

            if (@params["shadowNormalBias"] != null)
                light.shadowNormalBias = ParamCoercion.CoerceFloat(@params["shadowNormalBias"], light.shadowNormalBias);

            if (@params["cookieSize"] != null)
                light.cookieSize = ParamCoercion.CoerceFloat(@params["cookieSize"], light.cookieSize);

            if (@params["cullingMask"] != null)
                light.cullingMask = ParamCoercion.CoerceInt(@params["cullingMask"], light.cullingMask);

            if (@params["enabled"] != null)
                light.enabled = ParamCoercion.CoerceBool(@params["enabled"], light.enabled);
        }

        #endregion

        #region Ambient Lighting

        private static object AmbientSet(JObject @params)
        {
            Undo.RecordObject(RenderSettings.defaultReflectionMode == DefaultReflectionMode.Skybox ? (UnityEngine.Object)RenderSettings.skybox : null, "Set Ambient");

            if (@params["ambientMode"] != null)
            {
                string modeStr = ParamCoercion.CoerceString(@params["ambientMode"], "Skybox");
                if (Enum.TryParse<AmbientMode>(modeStr, true, out var mode))
                    RenderSettings.ambientMode = mode;
            }

            if (@params["ambientColor"] != null)
                RenderSettings.ambientLight = ParseColor(@params["ambientColor"], RenderSettings.ambientLight);

            if (@params["ambientSkyColor"] != null)
                RenderSettings.ambientSkyColor = ParseColor(@params["ambientSkyColor"], RenderSettings.ambientSkyColor);

            if (@params["ambientEquatorColor"] != null)
                RenderSettings.ambientEquatorColor = ParseColor(@params["ambientEquatorColor"], RenderSettings.ambientEquatorColor);

            if (@params["ambientGroundColor"] != null)
                RenderSettings.ambientGroundColor = ParseColor(@params["ambientGroundColor"], RenderSettings.ambientGroundColor);

            if (@params["ambientIntensity"] != null)
                RenderSettings.ambientIntensity = ParamCoercion.CoerceFloat(@params["ambientIntensity"], RenderSettings.ambientIntensity);

            EditorSceneManager.MarkAllScenesDirty();

            return new SuccessResponse("Ambient lighting updated.", new
            {
                ambientMode = RenderSettings.ambientMode.ToString(),
                ambientColor = ColorToHex(RenderSettings.ambientLight),
                ambientIntensity = RenderSettings.ambientIntensity
            });
        }

        private static object AmbientGet(JObject @params)
        {
            return new SuccessResponse("Ambient lighting settings.", new
            {
                ambientMode = RenderSettings.ambientMode.ToString(),
                ambientColor = ColorToHex(RenderSettings.ambientLight),
                ambientSkyColor = ColorToHex(RenderSettings.ambientSkyColor),
                ambientEquatorColor = ColorToHex(RenderSettings.ambientEquatorColor),
                ambientGroundColor = ColorToHex(RenderSettings.ambientGroundColor),
                ambientIntensity = RenderSettings.ambientIntensity
            });
        }

        #endregion

        #region Fog

        private static object FogSet(JObject @params)
        {
            if (@params["fogEnabled"] != null)
                RenderSettings.fog = ParamCoercion.CoerceBool(@params["fogEnabled"], RenderSettings.fog);

            if (@params["fogMode"] != null)
            {
                string modeStr = ParamCoercion.CoerceString(@params["fogMode"], "Linear");
                if (Enum.TryParse<FogMode>(modeStr, true, out var mode))
                    RenderSettings.fogMode = mode;
            }

            if (@params["fogColor"] != null)
                RenderSettings.fogColor = ParseColor(@params["fogColor"], RenderSettings.fogColor);

            if (@params["fogDensity"] != null)
                RenderSettings.fogDensity = ParamCoercion.CoerceFloat(@params["fogDensity"], RenderSettings.fogDensity);

            if (@params["fogStart"] != null)
                RenderSettings.fogStartDistance = ParamCoercion.CoerceFloat(@params["fogStart"], RenderSettings.fogStartDistance);

            if (@params["fogEnd"] != null)
                RenderSettings.fogEndDistance = ParamCoercion.CoerceFloat(@params["fogEnd"], RenderSettings.fogEndDistance);

            EditorSceneManager.MarkAllScenesDirty();

            return new SuccessResponse("Fog settings updated.", new
            {
                enabled = RenderSettings.fog,
                mode = RenderSettings.fogMode.ToString(),
                color = ColorToHex(RenderSettings.fogColor),
                density = RenderSettings.fogDensity,
                startDistance = RenderSettings.fogStartDistance,
                endDistance = RenderSettings.fogEndDistance
            });
        }

        private static object FogGet(JObject @params)
        {
            return new SuccessResponse("Fog settings.", new
            {
                enabled = RenderSettings.fog,
                mode = RenderSettings.fogMode.ToString(),
                color = ColorToHex(RenderSettings.fogColor),
                density = RenderSettings.fogDensity,
                startDistance = RenderSettings.fogStartDistance,
                endDistance = RenderSettings.fogEndDistance
            });
        }

        #endregion

        #region Reflection Probe

        private static object ReflectionProbeCreate(JObject @params)
        {
            Vector3 position = ParseVector3(@params["position"], Vector3.zero);
            Vector3 size = ParseVector3(@params["probeSize"], new Vector3(10, 10, 10));

            GameObject probeGo = new GameObject("Reflection Probe");
            probeGo.transform.position = position;

            ReflectionProbe probe = probeGo.AddComponent<ReflectionProbe>();
            probe.size = size;

            if (@params["probeResolution"] != null)
                probe.resolution = ParamCoercion.CoerceInt(@params["probeResolution"], 128);

            if (@params["probeHdr"] != null)
                probe.hdr = ParamCoercion.CoerceBool(@params["probeHdr"], true);

            if (@params["probeImportance"] != null)
                probe.importance = ParamCoercion.CoerceInt(@params["probeImportance"], 1);

            Undo.RegisterCreatedObjectUndo(probeGo, "Create Reflection Probe");
            EditorSceneManager.MarkSceneDirty(probeGo.scene);

            return new SuccessResponse("Reflection probe created.", new
            {
                instanceID = probeGo.GetInstanceID(),
                name = probeGo.name,
                position = new { x = position.x, y = position.y, z = position.z },
                size = new { x = size.x, y = size.y, z = size.z },
                resolution = probe.resolution
            });
        }

        private static object ReflectionProbeBake(JObject @params)
        {
            ReflectionProbe probe = FindReflectionProbe(@params);
            if (probe == null)
            {
                return new ErrorResponse("Reflection probe not found.");
            }

            // Trigger bake
            Lightmapping.BakeReflectionProbe(probe, probe.bakedTexture?.name ?? "ReflectionProbe");

            return new SuccessResponse($"Reflection probe '{probe.name}' baking started.", new
            {
                instanceID = probe.gameObject.GetInstanceID(),
                name = probe.name
            });
        }

        #endregion

        #region Lightmap Baking

        private static object LightmapBake(JObject @params)
        {
            string qualityStr = ParamCoercion.CoerceString(@params["bakeQuality"], "Medium");
            
            // Note: Lightmap baking is an async operation
            if (Lightmapping.isRunning)
            {
                return new ErrorResponse("Lightmap baking is already in progress.");
            }

            Lightmapping.BakeAsync();

            return new SuccessResponse("Lightmap baking started.", new
            {
                isRunning = Lightmapping.isRunning,
                quality = qualityStr
            });
        }

        private static object LightingGetSettings(JObject @params)
        {
            return new SuccessResponse("Current lighting settings.", new
            {
                ambient = new
                {
                    mode = RenderSettings.ambientMode.ToString(),
                    color = ColorToHex(RenderSettings.ambientLight),
                    intensity = RenderSettings.ambientIntensity
                },
                fog = new
                {
                    enabled = RenderSettings.fog,
                    mode = RenderSettings.fogMode.ToString(),
                    color = ColorToHex(RenderSettings.fogColor),
                    density = RenderSettings.fogDensity
                },
                lightmapping = new
                {
                    isRunning = Lightmapping.isRunning,
                    lightmapsCount = LightmapSettings.lightmaps?.Length ?? 0
                }
            });
        }

        #endregion

        #region Helpers

        private static Light FindLight(JObject @params)
        {
            JToken targetToken = @params["target"];
            if (targetToken == null) return null;

            if (targetToken.Type == JTokenType.Integer)
            {
                int instanceId = targetToken.Value<int>();
                GameObject go = GameObjectLookup.FindById(instanceId);
                return go?.GetComponent<Light>();
            }

            string targetStr = targetToken.ToString();
            if (int.TryParse(targetStr, out int parsedId))
            {
                GameObject go = GameObjectLookup.FindById(parsedId);
                if (go != null) return go.GetComponent<Light>();
            }

            GameObject found = GameObjectLookup.FindByTarget(targetToken, "by_name", true);
            return found?.GetComponent<Light>();
        }

        private static ReflectionProbe FindReflectionProbe(JObject @params)
        {
            JToken targetToken = @params["target"];
            if (targetToken == null)
            {
                return UnityEngine.Object.FindFirstObjectByType<ReflectionProbe>();
            }

            if (targetToken.Type == JTokenType.Integer)
            {
                int instanceId = targetToken.Value<int>();
                GameObject go = GameObjectLookup.FindById(instanceId);
                return go?.GetComponent<ReflectionProbe>();
            }

            string targetStr = targetToken.ToString();
            if (int.TryParse(targetStr, out int parsedId))
            {
                GameObject go = GameObjectLookup.FindById(parsedId);
                if (go != null) return go.GetComponent<ReflectionProbe>();
            }

            GameObject found = GameObjectLookup.FindByTarget(targetToken, "by_name", true);
            return found?.GetComponent<ReflectionProbe>();
        }

        private static object LightNotFoundError(JObject @params)
        {
            return new ErrorResponse($"Light not found: '{@params["target"]}'.");
        }

        private static Vector3 ParseVector3(JToken token, Vector3 defaultValue)
        {
            if (token == null) return defaultValue;

            if (token is JArray arr && arr.Count >= 3)
            {
                return new Vector3(
                    ParamCoercion.CoerceFloat(arr[0], defaultValue.x),
                    ParamCoercion.CoerceFloat(arr[1], defaultValue.y),
                    ParamCoercion.CoerceFloat(arr[2], defaultValue.z)
                );
            }

            if (token is JObject obj)
            {
                return new Vector3(
                    ParamCoercion.CoerceFloat(obj["x"], defaultValue.x),
                    ParamCoercion.CoerceFloat(obj["y"], defaultValue.y),
                    ParamCoercion.CoerceFloat(obj["z"], defaultValue.z)
                );
            }

            return defaultValue;
        }

        private static Color ParseColor(JToken token, Color defaultValue)
        {
            if (token == null) return defaultValue;

            // Hex string
            if (token.Type == JTokenType.String)
            {
                string hex = token.ToString();
                if (ColorUtility.TryParseHtmlString(hex, out Color color))
                    return color;
            }

            // Array [r, g, b] or [r, g, b, a]
            if (token is JArray arr && arr.Count >= 3)
            {
                return new Color(
                    ParamCoercion.CoerceFloat(arr[0], defaultValue.r),
                    ParamCoercion.CoerceFloat(arr[1], defaultValue.g),
                    ParamCoercion.CoerceFloat(arr[2], defaultValue.b),
                    arr.Count >= 4 ? ParamCoercion.CoerceFloat(arr[3], defaultValue.a) : 1f
                );
            }

            // Object {r, g, b, a}
            if (token is JObject obj)
            {
                return new Color(
                    ParamCoercion.CoerceFloat(obj["r"], defaultValue.r),
                    ParamCoercion.CoerceFloat(obj["g"], defaultValue.g),
                    ParamCoercion.CoerceFloat(obj["b"], defaultValue.b),
                    ParamCoercion.CoerceFloat(obj["a"], defaultValue.a)
                );
            }

            return defaultValue;
        }

        private static string ColorToHex(Color color)
        {
            return $"#{ColorUtility.ToHtmlStringRGBA(color)}";
        }

        #endregion
    }
}

File: Server/src/services/tools/manage_lighting.py
"""
Tool for managing Lighting in Unity.
Supports light creation, configuration, ambient settings, fog, and reflection probes.
"""
from typing import Annotated, Any, Literal

from fastmcp import Context
from services.registry import mcp_for_unity_tool
from services.tools import get_unity_instance_from_context
from transport.unity_transport import send_with_unity_instance
from transport.legacy.unity_connection import async_send_command_with_retry
from services.tools.preflight import preflight


@mcp_for_unity_tool(
    description="Manages Unity Lighting - create lights, configure properties, set ambient/fog, bake lightmaps, and manage reflection probes. Actions: light_create, light_configure, light_get_info, ambient_set, fog_set, fog_get, reflection_probe_create, reflection_probe_bake, lightmap_bake, lighting_get_settings."
)
async def manage_lighting(
    ctx: Context,
    action: Annotated[
        Literal[
            "light_create", "light_configure", "light_get_info",
            "ambient_set", "ambient_get",
            "fog_set", "fog_get",
            "reflection_probe_create", "reflection_probe_bake",
            "lightmap_bake", "lighting_get_settings"
        ],
        "Action to perform on lighting"
    ],
    target: Annotated[
        str | int,
        "Target Light/ReflectionProbe GameObject - instance ID or name/path"
    ] | None = None,
    # Light creation/configuration
    light_type: Annotated[
        Literal["Directional", "Point", "Spot", "Area"],
        "Type of light to create"
    ] | None = None,
    color: Annotated[
        list[float] | dict | str,
        "Light color [r, g, b] (0-1) or hex string"
    ] | None = None,
    intensity: Annotated[float, "Light intensity"] | None = None,
    range: Annotated[float, "Point/Spot light range"] | None = None,
    spot_angle: Annotated[float, "Spot light angle in degrees"] | None = None,
    inner_spot_angle: Annotated[float, "Spot light inner angle"] | None = None,
    shadows: Annotated[
        Literal["None", "Hard", "Soft"],
        "Shadow type"
    ] | None = None,
    shadow_strength: Annotated[float, "Shadow strength (0-1)"] | None = None,
    shadow_resolution: Annotated[
        Literal["FromQualitySettings", "Low", "Medium", "High", "VeryHigh"],
        "Shadow resolution"
    ] | None = None,
    shadow_bias: Annotated[float, "Shadow bias"] | None = None,
    shadow_normal_bias: Annotated[float, "Shadow normal bias"] | None = None,
    cookie: Annotated[str, "Path to cookie texture"] | None = None,
    cookie_size: Annotated[float, "Cookie projection size"] | None = None,
    culling_mask: Annotated[int | str, "Culling mask layers"] | None = None,
    position: Annotated[list[float] | dict, "Light position [x, y, z]"] | None = None,
    rotation: Annotated[list[float] | dict, "Light rotation [x, y, z]"] | None = None,
    # Ambient settings
    ambient_mode: Annotated[
        Literal["Skybox", "Trilight", "Flat", "Custom"],
        "Ambient lighting mode"
    ] | None = None,
    ambient_color: Annotated[
        list[float] | dict | str,
        "Ambient color for Flat mode"
    ] | None = None,
    ambient_sky_color: Annotated[list[float] | dict, "Sky color for Trilight"] | None = None,
    ambient_equator_color: Annotated[list[float] | dict, "Equator color for Trilight"] | None = None,
    ambient_ground_color: Annotated[list[float] | dict, "Ground color for Trilight"] | None = None,
    ambient_intensity: Annotated[float, "Ambient intensity multiplier"] | None = None,
    # Fog settings
    fog_enabled: Annotated[bool, "Enable/disable fog"] | None = None,
    fog_mode: Annotated[
        Literal["Linear", "Exponential", "ExponentialSquared"],
        "Fog mode"
    ] | None = None,
    fog_color: Annotated[list[float] | dict | str, "Fog color"] | None = None,
    fog_density: Annotated[float, "Fog density for exponential modes"] | None = None,
    fog_start: Annotated[float, "Fog start distance for linear mode"] | None = None,
    fog_end: Annotated[float, "Fog end distance for linear mode"] | None = None,
    # Reflection probe
    probe_resolution: Annotated[int, "Reflection probe resolution"] | None = None,
    probe_size: Annotated[list[float], "Probe box size [x, y, z]"] | None = None,
    probe_center: Annotated[list[float], "Probe center offset [x, y, z]"] | None = None,
    probe_hdr: Annotated[bool, "Enable HDR for probe"] | None = None,
    probe_importance: Annotated[int, "Probe importance for blending"] | None = None,
    # Lightmap baking
    bake_quality: Annotated[
        Literal["Low", "Medium", "High", "VeryHigh"],
        "Lightmap bake quality"
    ] | None = None,
    bake_mode: Annotated[
        Literal["Shadowmask", "Subtractive", "Baked"],
        "Mixed lighting mode"
    ] | None = None,
) -> dict[str, Any]:
    """
    Manage Unity Lighting.

    Actions:
    - light_create: Create a new light
    - light_configure: Configure light properties
    - light_get_info: Get light information
    - ambient_set: Set ambient lighting
    - ambient_get: Get ambient settings
    - fog_set: Configure fog
    - fog_get: Get fog settings
    - reflection_probe_create: Create reflection probe
    - reflection_probe_bake: Bake reflection probe
    - lightmap_bake: Bake lightmaps
    - lighting_get_settings: Get all lighting settings

    Examples:
    - Create light: action="light_create", light_type="Point", color=[1,0.9,0.8], intensity=2
    - Set fog: action="fog_set", fog_enabled=true, fog_mode="Linear", fog_start=10, fog_end=100
    - Bake lightmaps: action="lightmap_bake", bake_quality="High"
    """
    unity_instance = get_unity_instance_from_context(ctx)

    gate = await preflight(ctx, wait_for_no_compile=True, refresh_if_dirty=True)
    if gate is not None:
        return gate.model_dump()

    if not action:
        return {
            "success": False,
            "message": "Missing required parameter 'action'."
        }

    # Actions that don't require target
    no_target_actions = ["light_create", "ambient_set", "ambient_get", "fog_set", "fog_get", 
                         "reflection_probe_create", "lightmap_bake", "lighting_get_settings"]
    if action not in no_target_actions and not target:
        return {
            "success": False,
            "message": "Missing required parameter 'target'. Specify Light/Probe GameObject."
        }

    try:
        params: dict[str, Any] = {"action": action}
        
        if target is not None:
            params["target"] = target

        # Light properties
        if light_type is not None:
            params["lightType"] = light_type
        if color is not None:
            params["color"] = color
        if intensity is not None:
            params["intensity"] = intensity
        if range is not None:
            params["range"] = range
        if spot_angle is not None:
            params["spotAngle"] = spot_angle
        if inner_spot_angle is not None:
            params["innerSpotAngle"] = inner_spot_angle
        if shadows is not None:
            params["shadows"] = shadows
        if shadow_strength is not None:
            params["shadowStrength"] = shadow_strength
        if shadow_resolution is not None:
            params["shadowResolution"] = shadow_resolution
        if shadow_bias is not None:
            params["shadowBias"] = shadow_bias
        if shadow_normal_bias is not None:
            params["shadowNormalBias"] = shadow_normal_bias
        if cookie is not None:
            params["cookie"] = cookie
        if cookie_size is not None:
            params["cookieSize"] = cookie_size
        if culling_mask is not None:
            params["cullingMask"] = culling_mask
        if position is not None:
            params["position"] = position
        if rotation is not None:
            params["rotation"] = rotation

        # Ambient settings
        if ambient_mode is not None:
            params["ambientMode"] = ambient_mode
        if ambient_color is not None:
            params["ambientColor"] = ambient_color
        if ambient_sky_color is not None:
            params["ambientSkyColor"] = ambient_sky_color
        if ambient_equator_color is not None:
            params["ambientEquatorColor"] = ambient_equator_color
        if ambient_ground_color is not None:
            params["ambientGroundColor"] = ambient_ground_color
        if ambient_intensity is not None:
            params["ambientIntensity"] = ambient_intensity

        # Fog settings
        if fog_enabled is not None:
            params["fogEnabled"] = fog_enabled
        if fog_mode is not None:
            params["fogMode"] = fog_mode
        if fog_color is not None:
            params["fogColor"] = fog_color
        if fog_density is not None:
            params["fogDensity"] = fog_density
        if fog_start is not None:
            params["fogStart"] = fog_start
        if fog_end is not None:
            params["fogEnd"] = fog_end

        # Reflection probe
        if probe_resolution is not None:
            params["probeResolution"] = probe_resolution
        if probe_size is not None:
            params["probeSize"] = probe_size
        if probe_center is not None:
            params["probeCenter"] = probe_center
        if probe_hdr is not None:
            params["probeHdr"] = probe_hdr
        if probe_importance is not None:
            params["probeImportance"] = probe_importance

        # Lightmap baking
        if bake_quality is not None:
            params["bakeQuality"] = bake_quality
        if bake_mode is not None:
            params["bakeMode"] = bake_mode

        response = await send_with_unity_instance(
            async_send_command_with_retry,
            unity_instance,
            "manage_lighting",
            params,
        )

        if isinstance(response, dict) and response.get("success"):
            return {
                "success": True,
                "message": response.get("message", f"Lighting {action} successful."),
                "data": response.get("data")
            }
        return response if isinstance(response, dict) else {"success": False, "message": str(response)}

    except Exception as e:
        return {"success": False, "message": f"Error managing lighting: {e!s}"}
